### ISSUE TITLE
Hang/crash when after a *lot* of consecutive requests

### ISSUE THREAD

**Issue by jschwartzenberg** [versat/cntlm#18]
I'm running into a situation where a lot of consecutive requests are causing cntlm to hang. The output ends with a huge amount of messages like:
```
[Thread 0x7ffff465b700 (LWP 46761) exited]
Joined thread 140736274474752; rc: 0
Joined thread 140736277956352; rc: 0
Joined thread 140736269600512; rc: 0
Joined thread 140737288541952; rc: 0
Joined thread 140737288333056; rc: 0
Joined thread 140737291048704; rc: 0
Joined thread 140737293137664; rc: 0
Joined thread 140736271759104; rc: 0
Joined thread 140736270853888; rc: 0
Joined thread 140737293764352; rc: 0
Joined thread 140736274753280; rc: 0
Joined thread 140737299195648; rc: 0
Joined thread 140737293694720; rc: 0
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4738700 (LWP 46762) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb75a3700 (LWP 46763) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff468e700 (LWP 46764) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff45b1700 (LWP 46765) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb7e12700 (LWP 46767) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb765e700 (LWP 46766) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb761a700 (LWP 46769) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff43c4700 (LWP 46768) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb75e7700 (LWP 46770) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4072700 (LWP 46771) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4eca700 (LWP 46772) exited]
Joined thread 140737294599936; rc: 0
Joined thread 140736269530880; rc: 0
Joined thread 140737293903616; rc: 0
Joined thread 140737292998400; rc: 0
Joined thread 140736278374144; rc: 0
Joined thread 140736270296832; rc: 0
Joined thread 140736270018304; rc: 0
Joined thread 140737290979072; rc: 0
Joined thread 140736269809408; rc: 0
Joined thread 140737287497472; rc: 0
Joined thread 140737302537984; rc: 0
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4e53700 (LWP 46778) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4c66700 (LWP 46777) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4a79700 (LWP 46776) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4bcd700 (LWP 46779) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff7f3c700 (LWP 46780) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4b89700 (LWP 46773) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4c11700 (LWP 46775) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4061700 (LWP 46782) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4d10700 (LWP 46781) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4c99700 (LWP 46784) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff476b700 (LWP 46785) exited]
Joined thread 140737302050560; rc: 0
Joined thread 140737300031232; rc: 0
Joined thread 140737298011904; rc: 0
Joined thread 140737299404544; rc: 0
Joined thread 140737353336576; rc: 0
Joined thread 140737299126016; rc: 0
Joined thread 140737299683072; rc: 0
Joined thread 140737287427840; rc: 0
Joined thread 140737300727552; rc: 0
Joined thread 140737300240128; rc: 0
Joined thread 140737294808832; rc: 0
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4d76700 (LWP 46786) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff423d700 (LWP 46793) exited]
proxy_thread: request rc = 0xffffffffffffffff
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4dcb700 (LWP 46794) exited]
[Thread 0x7ffff4f52700 (LWP 46795) exited]
proxy_thread: request rc = 0xffffffffffffffff
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4749700 (LWP 46797) exited]
[Thread 0x7ffff46c1700 (LWP 46796) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4d21700 (LWP 46798) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4e20700 (LWP 46799) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff41c6700 (LWP 46800) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4e42700 (LWP 46802) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4cdd700 (LWP 46801) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4d98700 (LWP 46787) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4ffc700 (LWP 46803) exited]
Joined thread 140737301145344; rc: 0
Joined thread 140737289377536; rc: 0
Joined thread 140737301493504; rc: 0
Joined thread 140737303095040; rc: 0
Joined thread 140737294669568; rc: 0
Joined thread 140737294112512; rc: 0
Joined thread 140737300797184; rc: 0
Joined thread 140737301841664; rc: 0
Joined thread 140737288890112; rc: 0
Joined thread 140737301980928; rc: 0
Joined thread 140737300518656; rc: 0
Joined thread 140737301284608; rc: 0
Joined thread 140737303791360; rc: 0
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4efd700 (LWP 46805) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4edb700 (LWP 46806) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4eec700 (LWP 46807) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff523e700 (LWP 46810) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4da9700 (LWP 46808) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4f74700 (LWP 46811) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff53f8700 (LWP 46812) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4d54700 (LWP 46814) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4cff700 (LWP 46815) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff7f4d700 (LWP 46816) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff420a700 (LWP 46804) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff4094700 (LWP 46817) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff402e700 (LWP 46818) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff5150700 (LWP 46819) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7ffff500d700 (LWP 46820) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb7807700 (LWP 46821) exited]
proxy_thread: request rc = 0xffffffffffffffff
[Thread 0x7fffb75c5700 (LWP 46822) exited]
Joined thread 140737302746880; rc: 0
Joined thread 140737302607616; rc: 0
Joined thread 140737302677248; rc: 0
Joined thread 140737306158848; rc: 0
Joined thread 140737301354240; rc: 0
Joined thread 140737303234304; rc: 0
Joined thread 140737307969280; rc: 0
Joined thread 140737301006080; rc: 0
Joined thread 140737300657920; rc: 0
Joined thread 140737353406208; rc: 0
Joined thread 140737289168640; rc: 0
Joined thread 140737287636736; rc: 0
Joined thread 140737287218944; rc: 0
Joined thread 140737305184000; rc: 0
Joined thread 140737303860992; rc: 0
Joined thread 140736272037632; rc: 0
Joined thread 140736269670144; rc: 0
```

I suspect something is going wrong with the thread counting or the mutex.

**Comment by jschwartzenberg** [comment]
I added crash, because I've seen that happening too, but now it seems one particular thread just hung.

**Comment by jschwartzenberg** [comment]
This is indeed resolved with #52!
